### PR TITLE
Refactored ModuleManager

### LINKED ISSUE #338: Replace PR "DropCommunicatorUtil is now a Module"
After PR #327 is merged, #326 will be redone based on the new master.

### PR DESCRIPTION
**Pull request by L-Henke**
This is based on and replaces #326.
Closes #338.

Descriptions copied from commit messages:

```
Refactored ModuleManager

Removed (currently) unneeded methods.

Added a wrapping method for Modules to register for Events. Its purpose
is to be able to enforce unique event name restrictions and event
disallow registering to certain events if required. This only disables a
direct way to register and doesn't prevent Modules from doing unwanted
things by Reflections or other approaches.
```

```
Refactored Module Class

Module class now doesn't get a EventEmitter in the constructor. Instead,
it can receive the required EventEmitter from its ModuleManager by a
package private method. This disallows direct access to the EventEmitter
for inheritors from Module which usually should not be in the module
package.

The required on() and emit() methods are wrapped for inheritors.
```

### CODE REVIEW

File: src/main/java/de/qabel/core/module/ModuleManager.java [hunk 1]
```diff
@@ -1,32 +1,28 @@
...
-	    @Override
-	    public void addURL(URL url) {
-	        super.addURL(url);
-	    }
-	}
+	private final EventEmitter eventEmitter;
+	private final ConfigActor configActor;
+	private final ContactsActor contactsActor;
+
+	private HashMap<Module, ModuleThread> modules;
```
Comment (cburkert) [inline src/main/java/de/qabel/core/module/ModuleManager.java]: Couldn't this be static as well?

Comment (cburkert) [inline src/main/java/de/qabel/core/module/ModuleManager.java]: ... I mean final.

File: src/main/java/de/qabel/core/module/ModuleManager.java [hunk 2]
```diff
@@ -87,50 +75,57 @@ public void run() {
 	 * @throws InstantiationException
 	 * @throws IllegalAccessException
 	 */
-	public void startModule(Class<?> module) throws InstantiationException, IllegalAccessException {
-		Module m = (Module) module.newInstance();
-		m.setModuleManager(this);
+	public <T extends Module> T startModule(Class<T> module) throws InstantiationException, IllegalAccessException, NoSuchMethodException, InvocationTargetException {
```
Comment (cburkert) [inline src/main/java/de/qabel/core/module/ModuleManager.java]: Why not use `<? extends Module>`? You don't need to know the concrete subclass, do you?

File: src/main/java/de/qabel/core/module/ModuleManager.java [hunk 3]
```diff
@@ -87,50 +75,57 @@ public void run() {
...
+			try {
+				startModule(cls.asSubclass(Module.class));
+			}
+			catch (ClassCastException e) {
+				throw new InstantiationError(className + " not a subclass of Module!");
+			}
+		} catch (Exception e) {
 			e.printStackTrace();
-			// Should not happen!
 			System.exit(1);
```
Comment (cburkert) [inline src/main/java/de/qabel/core/module/ModuleManager.java]: Do we really want to exit here? I'd throw a runtime exception.
(I know you didn't write this, but at least you removed the line comment ;-))

File: src/main/java/de/qabel/core/module/ModuleManager.java [hunk 4]
```diff
@@ -87,50 +75,57 @@ public void run() {
...
-	private DropActor dropActor;
+	/**
+	 * Wraps the EventListener on method. Allows to
+	 * force unique namespaces for modules.
+	 * @param event Event name to register for
+	 * @param module Module to add as EventListener
+	 * @return True if successfully registered
+	 */
+	public boolean on(String event, Module module) {
+		if (true) { //TODO: This could enforce unique names and disallow registering to certain events.
```
Comment (cburkert) [inline src/main/java/de/qabel/core/module/ModuleManager.java]: Is there an use for that todo?

**Comment by thechauffeur** [comment]
Wait with merge until Qabel/qabel-helloworld-module#16 and Qabel/qabel-desktop#16 look like something we want (i.e. how we want a module and an app to interact with each other).

Of course also wait until this is properly reviewed :)

File: src/main/java/de/qabel/core/module/ModuleManager.java (hunk 2 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/module/ModuleManager.java]: Forget it. Your approach is better.

File: src/main/java/de/qabel/core/module/ModuleManager.java (hunk 4 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/module/ModuleManager.java]: Typo hell, I meant issue, sorry.

Comment (L-Henke) [inline src/main/java/de/qabel/core/module/ModuleManager.java]: Now it is: #354

**Pull request merged**